backend/src/auth/mod.rs

```diff
@@ -135,10 +152,34 @@ pub(crate) struct User {
 }
 
 /// Returns a representation of the optional username useful for logging.
-pub(crate) fn debug_log_username(session: &Option<User>) -> Cow<'static, str> {
+pub(crate) fn debug_log_username(session: &AuthContext) -> Cow<'static, str> {
+    use AuthContext::*;
     match session {
-        None => "none".into(),
-        Some(user) => format!("'{}'", user.username).into(),
+        Anonymous => "none".into(),
+        TrustedExternal => "trusted external".into(),
+        User(user) => format!("'{}'", user.username).into(),
+    }
+}
+
+impl AuthContext {
+    pub(crate) async fn new(
+        headers: &HeaderMap,
+        auth_config: &AuthConfig,
+        db: &Client,
+    ) -> Result<Self, PgError> {
+
+        if let Some(request_trusted_key) = headers.get("x-tobira-trusted-key") {
+            if let Some(ref config_trusted_key) = auth_config.trusted_key {
+                if config_trusted_key.expose_secret() == request_trusted_key {
+                    return Ok(Self::TrustedExternal);
+                }
+            }
+        }
```

Comment:
I've thought a bit about this as it seems one can write this much shorter. And one kind of can with `map_or` and `zip` or `matches!`. But in practice, I feel none of the other versions improve upon this. So I think it's fine just leaving it as is. I think only one version I would also consider fine:

```rust
        if auth_config.trusted_key.is_some() {
            if auth_config.trusted_key == headers.get("x-tobira-trusted-key") {
                return Ok(Self::TrustedExternal);        
            }
        }
```

But if you leave the current solution, remove `ref` (see other comment about this):

```suggestion
        if let Some(request_trusted_key) = headers.get("x-tobira-trusted-key") {
            if let Some(config_trusted_key) = &auth_config.trusted_key {
                if config_trusted_key.expose_secret() == request_trusted_key {
                    return Ok(Self::TrustedExternal);
                }
            }
        }
```

Reply:
Your suggestion doesn't work as is, because it tries to compare an `Option<Secret<String>>` with a an `Option<&HeaderValue>`, so you would have to `unwrap` or `map` the first while you already know that there is a `Some` in there. This is always kind of a red flag to me. So I would say let's keep it? I'll fix the `ref` thing, though.